R/html-tweak.R

```diff
@@ -251,38 +251,43 @@ tweak_homepage_html <- function(html, strip_header = FALSE) {
 
 # Mutates `html`, removing the badge container
 badges_extract <- function(html) {
-  # First try specially named div;
+  # First try specially named element;
   x <- xml2::xml_find_first(html, "//div[@id='badges']")
+  strict <- FALSE
 
   # then try usethis-readme-like paragraph;
   if (length(x) == 0) {
-    x <- xml2::xml_find_all(html, ".//*/comment()[contains(., 'badges: start')]/following-sibling::p[1]")
+    # Find start comment, then all elements after
+    # which are followed by the end comment.
+    x <- xml2::xml_find_all(html, "
+      //comment()[contains(., 'badges: start')][1]
+      /following-sibling::*[following-sibling::comment()[contains(., 'badges: end')]]
+    ")
   }
 
   # finally try first paragraph
   if (length(x) == 0) {
     x <- xml2::xml_find_first(html, "//p")
+    strict <- TRUE
   }
 
   # No paragraph
   if (length(x) == 0) {
     return(character())
   }
 
-  # No non-whitespace characters outside of tags
-  if (xml2::xml_text(x, trim = TRUE) != "") {
+  # If we guessed the element,
+  # we only proceed if there is no text
+  if (strict && any(xml2::xml_text(x, trim = TRUE) != "")) {
     return(character())
   }
 
-  badges <- xml2::xml_children(x)
+  # Proceed if we find image-containing links
+  badges <- xml2::xml_find_all(x, "//a[img]")
```

Comment:
@flying-sheep This is likely the cause of #1255.

Reply:
Should be `.//a[img]`?

Reply:
That fixes the issue, but then the associated test fails.

Reply:
@flying-sheep this is the failing test, I'm not sure what case it's testing for (I don't know what you mean by "special element"):

``` r
test_that("badges in special element can be accompanied by text", {
  expect_equal(
    badges_extract_text('<p></p><p><span id="badges"><a href="x"><img src="y"></a>Hi!</span></p>'),
    '<a href="x"><img src="y"></a>'
  )
})

```

Reply:
Special element = element with `id="badges"`. Sorry about the bug! I don’t understand why the test fails with the `.//` variant, as manually executing the same code works?!

```r
> pkgdown:::badges_extract_text('<p></p><div id="badges"><a href="x"><img src="y"></a></div>')
[1] "<a href=\"x\"><img src=\"y\"></a>"
> testthat::...()
[…]
test-html-tweak.R:173: failure: badges in special element can be accompanied by text
badges_extract_text("<p></p><p><span id=\"badges\"><a href=\"x\"><img src=\"y\"></a>Hi!</span></p>") not equal to "<a href=\"x\"><img src=\"y\"></a>".
Lengths differ: 0 is not 1
[…]
```

Reply:
See  #1258